### ISSUE TITLE
Bug: [no-unsafe-assignment] Misleading error message for object destructuring

### ISSUE THREAD

**Issue by szhsin** [typescript-eslint/typescript-eslint#11387]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.8.2&fileType=.tsx&code=MYewdgzgLgBAZiEMC8MA8AVANAPgBQAmAhlEQFwwYCUKOMeA3jMaVjAJYQCiAtgA5QAnhQCELIjAC%2BVANwAoUJFhNxUlPER4AUgGUA8gDkAdHyIAnCAFM8AcgaSbVWUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9kySgHMmAW0Qt0kRNGgdokcGAC%2BITUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const foo = <T,>(data: T) => ({ data, isEmpty: !data });

const { data } = foo(JSON.parse('{}'));
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unsafe-assignment": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

If the linter wants to report an unsafe destructuring of `any` type, the message should clearly indicate it’s due to destructuring from an object, not an array or tuple.

### Actual Result

The code produces the following error:

```
Unsafe array destructuring of a tuple element with an `any` value
```

`data` is inferred as `any`, which makes sense and it's fair for the rule to report that.

What doesn't make sense is the error message — it's an object destructuring, not an array. The message is confusing and makes it sound like a tuple or array is involved, when it's clearly just pulling a property off an object.

**Comment by divyanshudg-git** [comment]
“I'd like to work on this. Assign me?”
